pallets/oracle/src/dia.rs

```diff
@@ -0,0 +1,75 @@
+use dia_oracle::DiaOracle;
+use orml_oracle::{DataProviderExtended, TimestampedValue};
+pub use primitives::{oracle::Key as OracleKey, CurrencyId, TruncateFixedPointToInt};
+use sp_std::marker;
+
+use sp_runtime::traits::Convert;
+
+struct X;
+
+impl Convert<OracleKey, Option<(Vec<u8>, Vec<u8>)>> for X {
+	fn convert(a: OracleKey) -> Option<(Vec<u8>, Vec<u8>)> {
+		todo!()
+	}
+}
+
+impl Convert<(Vec<u8>, Vec<u8>), Option<OracleKey>> for X {
+	fn convert(a: (Vec<u8>, Vec<u8>)) -> Option<OracleKey> {
+		todo!()
+	}
+}
+
+pub struct DiaOracleAdapter<
+	DiaPallet: DiaOracle,
+	UnsignedFixedPoint,
+	Moment,
+	ConvertKey,
+	ConvertPrice,
+	ConvertMoment,
```

Comment:
If we need all three I would group them under a new trait and name it something like eg. `TypeConversion`.

```
trait TypeConversion<TargetKey, TargetPrice, TargetMoment> {
   type ConvertKey: Convert<TargetKey, Option<(Vec<u8>, Vec<u8>)>> + Convert<(Vec<u8>, Vec<u8>), Option<TargetKey>>,
   type ConvertPrice:Convert<TargetPrice, Option<u128>> + Convert<u128, Option<TargetPrice>>,
   type ConvertMoment: Convert<TargetMoment, Option<u64>> + Convert<u64, Option<TargetMoment>>,
>
```

then in the `where` statement of the `impl` block you just specify the target types. The source types should always be the same and we don't need to make them generic since they are set by DIA.

Reply:
Thanks. Good idea! 
I just take as example [fungibles_adapter.rs](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-builder/src/fungibles_adapter.rs#L63-L72) from xcm-builder

There are two different type without combination in one trait.
```
ConvertAssetId: Convert<MultiLocation, AssetId>,
ConvertBalance: Convert<u128, Balance>,
```

Do we need introduce one more trait `TypeConversion` to group them together?
@ebma

Reply:
I think it would be better to group it into another trait since we don't have just two but three. But that's just my personal opinion. Would you prefer to keep the three conversions separated?

Reply:
For now i would like to keep them separated. Try to mock in mock.rs and then complicate and introduce type to combine them together.
Because now there is a problem more complicated: `fn _set_exchange_rate`
https://github.com/pendulum-chain/spacewalk/blob/dev/milestone-3/pallets/oracle/src/lib.rs#L392-L401
Used by other pallet and benchmarking..
What is the best way to solve and replace it?
Because with new oracle-extension implementation there is no way to update storage...

Reply:
Hmm, that's problematic indeed. How about we make our `DiaOracleAdapter` not only implement the DataProviderExtended trait but also the [DataFeeder](https://docs.rs/orml-traits/0.4.0/orml_traits/data_provider/trait.DataFeeder.html) trait? Then we can change this `_set_exchange_rate()` function to call the [feed_value](https://docs.rs/orml-traits/0.4.0/orml_traits/data_provider/trait.DataFeeder.html#tymethod.feed_value) function.

Reply:
yes. we can. but we will not use datafeed trait in prodaction. only for testing. and this trait bound will be redundant for prodaction runtime.

Reply:
How about we only require the additional `DataFeeder` trait bound when the `testing-utils` feature is set on the `oracle` crate then? The `_set_exchange_rate` function is also only available if that feature is set